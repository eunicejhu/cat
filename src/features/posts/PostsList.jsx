import React from "react";
import { useSelector } from "react-redux";
import AddPostForm from "./AddPostForm.tsx";

const PostsList = () => {
  const posts = useSelector((state) => state.posts);
  const renderPosts =
    posts &&
    posts.map((post) => (
      <article className="post-excerpt" key={post.id}>
        <h3>{post.title}</h3>
        <p className="post-content">{post.content.substring(0, 100)}</p>
      </article>
    ));

  return (
    <>
      <section className="add-post-form">
        <AddPostForm />
      </section>
      <section className="posts-list">
        <h2>Posts</h2>
        {posts && posts.length ? renderPosts : "No posts"}
      </section>
    </>
  );
};

export default PostsList;
