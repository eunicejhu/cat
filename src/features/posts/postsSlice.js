import { createSlice } from "@reduxjs/toolkit";
import initialState from "../../helpers/initialState";

const postsSlice = createSlice({
  name: "posts",
  initialState: initialState.posts,
  reducers: {},
});

export default postsSlice.reducer;
