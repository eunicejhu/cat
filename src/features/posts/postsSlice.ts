import { createSlice, nanoid } from "@reduxjs/toolkit";

export type State = Post[];
export interface Post {
  id: string;
  title: string;
  content: string;
}

export interface PostAddedAction {
  type: string;
  payload: Post;
}

export interface PostUpdatedAction {
  type: string;
  payload: Post;
}

const postsSlice = createSlice({
  name: "posts",
  initialState: [],
  reducers: {
    postAdded: {
      reducer(state: State, action: PostAddedAction) {
        state.push(action.payload);
      },
      prepare({ title, content }) {
        return { payload: { id: nanoid(), title, content } };
      },
    },
    postUpdated: (state: State, action: PostUpdatedAction) => {
      const { id, title, content } = action.payload;
      const existingPost = state.find((post) => post.id === id);
      if (existingPost) {
        existingPost.title = title;
        existingPost.content = content;
      }
    },
  },
});
export const { postAdded, postUpdated } = postsSlice.actions;
export default postsSlice.reducer;
