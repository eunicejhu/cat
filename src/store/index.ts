import { configureStore } from "@reduxjs/toolkit";
import postsReducer, { Post } from "../features/posts/postsSlice";
import usersReducer from "../features/users/usersSlice";
import notificationsReducer, {
  Notification,
} from "../features/notifications/notificationsSlice";
import { useDispatch } from "react-redux";

const store = configureStore({
  reducer: {
    posts: postsReducer,
    users: usersReducer,
    notifications: notificationsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();

export type State = ReturnType<typeof store.getState>;
// selectors
export const selectAllPosts = (state: State): Post[] => state.posts.data;
export const selectPostById = (
  state: State,
  postId: string
): Post | undefined => state.posts.data.find((post) => post.id === postId);
export const selectAllNotifications = (state: State): Notification[] =>
  state.notifications;

export default store;
