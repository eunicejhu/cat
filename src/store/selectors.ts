// selectors
import { State } from "./index";
import { Notification } from "../features/notifications/notificationsSlice";
import { Post } from "../features/posts/postsSlice";
export const selectAllPosts = (state: State): Post[] => state.posts.data;
export const selectPostById = (
  state: State,
  postId: string
): Post | undefined => state.posts.data.find((post) => post.id === postId);
export const selectAllNotifications = (state: State): Notification[] =>
  state.notifications;
