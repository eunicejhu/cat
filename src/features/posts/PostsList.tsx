import React from "react";
import { useSelector } from "react-redux";
import { useRouteMatch, Link } from "react-router-dom";
import { isAfter, parseISO } from "date-fns";
import AddPostForm from "./AddPostForm";
import { State } from "../../store";
import PostAuthor from "./PostAuthor";
import TimeAgo from "./TimeAgo";
const PostsList = () => {
  const { path } = useRouteMatch();
  const posts = useSelector((state: State) => state.posts);
  const sortedPosts =
    posts && posts.slice().sort((a, b) => a.date.localeCompare(b.date));
  const renderPosts =
    sortedPosts &&
    sortedPosts.map((post) => (
      <article className="post-excerpt" key={post.id}>
        <h3>{post.title}</h3>
        <p className="post-content">{post.content.substring(0, 100)}</p>
        <PostAuthor userId={post.userId} />
        <TimeAgo date={post.date} />
        <p>
          <Link to={`${path}posts/${post.id}`}>See more</Link>
        </p>
      </article>
    ));

  return (
    <>
      <section className="add-post-form">
        <AddPostForm />
      </section>
      <section className="posts-list">
        <h2>Posts</h2>
        {posts && posts.length ? renderPosts : "No posts"}
      </section>
    </>
  );
};

export default PostsList;
