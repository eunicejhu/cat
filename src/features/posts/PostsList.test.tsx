import React from "react";
import { fireEvent, render } from "@testing-library/react";
import PostsList from "./PostsList";
import { MemoryRouter, Route, BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../../store/index";
import Client from "../../api/client";
import { INITIAL_STATE } from "../../test/mock_data";
import { fetchPosts } from "./postsSlice";

jest.mock("../../api/client");

describe("PostsList test:", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test("show initial postsList", async () => {
    Client.fetchPost = jest
      .fn()
      .mockResolvedValueOnce({ data: INITIAL_STATE.posts.data });

    const ui = (
      <MemoryRouter initialEntries={["/posts"]} initialIndex={0}>
        <Provider store={store}>
          <Route exact path="/posts">
            <PostsList />
          </Route>
        </Provider>
      </MemoryRouter>
    );
    const { container, findByText } = render(ui);
    expect(await findByText(/First test Post!/i)).toBeInTheDocument();
    expect(store.getState().posts.data.length).toBe(2);

    expect(container.querySelectorAll(".post-excerpt").length).toBe(2);
  });

  test("click Seemore of first post direct to SinglePostPage", async () => {
    Client.fetchPost = jest
      .fn()
      .mockResolvedValueOnce({ data: INITIAL_STATE.posts.data });
    window.history.pushState({}, "Posts List", "/");
    const ui = (
      <BrowserRouter>
        <Provider store={store}>
          <Route exact path="/">
            <PostsList />
          </Route>
        </Provider>
      </BrowserRouter>
    );
    const { findAllByText } = render(ui);
    const firstPostSeeMoreButton = (await findAllByText(/See more/i))[0];
    fireEvent.click(firstPostSeeMoreButton);
    expect(window.location.pathname).toBe("/posts/1");
  });
  test("show no posts when posts is []", async () => {
    Client.fetchPost = jest.fn().mockResolvedValueOnce({ data: [] });
    const ui = (
      <MemoryRouter initialEntries={["/posts"]} initialIndex={0}>
        <Provider store={store}>
          <Route exact path="/posts">
            <PostsList />
          </Route>
        </Provider>
      </MemoryRouter>
    );
    const { findByText } = render(ui);
    store.dispatch(fetchPosts());
    expect(await findByText(/No Post/i)).toBeInTheDocument();
    expect(store.getState().posts.data.length).toBe(0);
  });
});
