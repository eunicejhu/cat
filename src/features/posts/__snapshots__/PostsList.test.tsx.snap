// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
Array [
  <section
    className="add-post-form"
  >
    <div
      className="add-post"
    >
      <form>
        <div
          className="field"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            data-testid="title"
            id="title"
            name="title"
            onChange={[Function]}
            value=""
            width={20}
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="content"
          >
            Content
          </label>
          <textarea
            data-testid="content"
            id="content"
            name="content"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          className="field"
        >
          <label
            htmlFor="users"
          >
            Author
          </label>
          <select
            data-testid="users"
            id="users"
            name="users"
            onChange={[Function]}
            value=""
          >
            <option
              value=""
            />
            <option
              value="0"
            >
              Tianna Jenkins
            </option>
            <option
              value="1"
            >
              Kevin Grant
            </option>
            <option
              value="2"
            >
              Madison Price
            </option>
          </select>
        </div>
        <div
          className="field"
        >
          <button
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Add Post
          </button>
        </div>
      </form>
    </div>
  </section>,
  <section
    className="posts-list"
  >
    <h2>
      Posts
    </h2>
  </section>,
]
`;
